File: lua/matching-brackets/matching-brackets.lua
return {
  valid = function(s)

    -- Hold key/value pair for terminal/inital bracket pairs respectively
    local complement = { [']'] = '[', [')'] = '(', ['}'] = '{' }

    -- Table will act like a stack and hold previously seen brackets
    local seen_brackets = {}

    -- Iterate through input string "s" for bracket characters
    local bracket_patterns = '[%[%](){}]'
    for bracket, _ in string.gmatch(s, bracket_patterns) do

      -- If the current bracket is a terminal bracket and matches with the last
      -- bracket in the seen_brackets stack, pop the last bracket from the stack
      if complement[bracket] and seen_brackets[#seen_brackets] == complement[bracket] then
        table.remove(seen_brackets)
      else
        table.insert(seen_brackets, bracket)
      end
    end

    -- A valid string should result in an empty table
    return #seen_brackets == 0

  end,
}

File: typescript/matching-brackets/matching-brackets.ts
const isBracket = (char: string): boolean => /[\(\)\{\}\[\]]/.test(char);

const complement: Map<string, string> = new Map<string, string>([
  ["[", "]"],
  ["{", "}"],
  ["(", ")"],
]);

export function isPaired(input: string): boolean {
  const stack: Array<string> = [];
  for (const char of input) {
    if (!isBracket(char)) {
      continue;
    }

    stack.push(char);
    if (stack.length < 2) {
      continue;
    }

    if (stack[stack.length - 1] === complement.get(stack[stack.length - 2])) {
      stack.pop();
      stack.pop();
    }
  }
  return stack.length === 0;
}
